controllers/provider-gcp/pkg/validator/shoot_validator.go

```diff
@@ -0,0 +1,157 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validator
+
+import (
+	"context"
+	"errors"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	gcpvalidation "github.com/gardener/gardener-extensions/controllers/provider-gcp/pkg/apis/gcp/validation"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	"github.com/gardener/gardener/pkg/apis/garden"
+)
+
+// getAllowedRegionZonesFromCloudProfile fetches the set of allowed zones from the Cloud Profile.
+func getAllowedRegionZonesFromCloudProfile(shoot *garden.Shoot, cloudProfile *gardencorev1beta1.CloudProfile) sets.String {
+	shootRegion := shoot.Spec.Region
+	for _, region := range cloudProfile.Spec.Regions {
+		if region.Name == shootRegion {
+			gcpZones := sets.NewString()
+			for _, gcpZone := range region.Zones {
+				gcpZones.Insert(gcpZone.Name)
+			}
+			return gcpZones
+		}
+	}
+	return nil
+}
+
+func (v *Shoot) validateShoot(ctx context.Context, shoot *garden.Shoot) error {
+	// Network validation
+	networkPath := field.NewPath("spec", "networking")
+
+	if errList := gcpvalidation.ValidateNetworking(shoot.Spec.Networking, networkPath); len(errList) != 0 {
+		return errList.ToAggregate()
+	}
+
+	// Provider validation
+	fldPath := field.NewPath("spec", "provider")
+
+	// InfrastructureConfig
+	infraConfigFldPath := fldPath.Child("infrastructureConfig")
+
+	if shoot.Spec.Provider.InfrastructureConfig == nil {
+		return field.Required(infraConfigFldPath, "InfrastructureConfig must be set for GCP shoots")
+	}
+
+	infraConfig, err := decodeInfrastructureConfig(v.decoder, shoot.Spec.Provider.InfrastructureConfig, infraConfigFldPath)
+	if err != nil {
+		return err
+	}
+
+	if errList := gcpvalidation.ValidateInfrastructureConfig(infraConfig, shoot.Spec.Networking.Nodes, shoot.Spec.Networking.Pods, shoot.Spec.Networking.Services); len(errList) != 0 {
+		return errList.ToAggregate()
+	}
+
+	cloudProfile := &gardencorev1beta1.CloudProfile{}
+	if err := v.client.Get(ctx, kutil.Key(shoot.Spec.CloudProfileName), cloudProfile); err != nil {
+		return err
+	}
+
+	controlPlaneConfigPath := fldPath.Child("controlPlaneConfig")
+	if shoot.Spec.Provider.ControlPlaneConfig != nil {
```

Comment:
Have you double checked that this works for all K8s versions Gardener supports, i.e. the region is not passed to CCM?

Reply:
It doesn't work, we looked into the CCM code. The region is needed to prevent calls against the GCE metadata server. Please ensure that a control plane config (and the region) are provided.